src/client_state.cpp

```diff
@@ -330,42 +331,119 @@ void wsrep::client_state::disable_streaming()
 //                                 TOI                                      //
 //////////////////////////////////////////////////////////////////////////////
 
-void wsrep::client_state::enter_toi_common()
+enum wsrep::provider::status
+wsrep::client_state::poll_enter_toi(
+    wsrep::unique_lock<wsrep::mutex>& lock,
+    const wsrep::key_array& keys,
+    const wsrep::const_buffer& buffer,
+    wsrep::ws_meta& meta,
+    int flags,
+    std::chrono::time_point<wsrep::clock> wait_until,
+    bool& timed_out)
 {
-    wsrep::unique_lock<wsrep::mutex> lock(mutex_);
+    WSREP_LOG_DEBUG(debug_log_level(),
+                    wsrep::log::debug_level_client_state,
+                    "poll_enter_toi: "
+                    << flags
+                    << ","
+                    << wait_until.time_since_epoch().count());
+    enum wsrep::provider::status status;
+    timed_out = false;
+    do
+    {
+        lock.unlock();
+        status = provider().enter_toi(id_, keys, buffer, meta, flags);
+        if (status != wsrep::provider::success &&
+            not toi_meta_.gtid().is_undefined())
+        {
+            // Successfully entered TOI, but the provider reported failure.
+            // This may happen for example if certification fails.
+            // Leave TOI before proceeding.
+            if (provider().leave_toi(id_, wsrep::mutable_buffer()))
+            {
+                wsrep::log_warning()
+                    << "Failed to leave TOI after failure in "
+                    << "poll_enter_toi()";
+            }
+            toi_meta_ = wsrep::ws_meta();
+        }
+        if (status == wsrep::provider::error_certification_failed ||
+            status == wsrep::provider::error_connection_failed)
+        {
+            ::usleep(100000);
```

Comment:
100msec sleep could make the loop too tight as `enter_toi()` requires communication over network.